### PR TITLE
Support serve + chat + generate with Mixtral teacher model 

### LINKED ISSUE #29: chat template is hard-coded
need to make the chat template model dependent as `lab generate` can use different models

### LINKED ISSUE #973: `ilab` serve and config (init) code should be model-dependent
there are parts of the model `serve` and `config` code that are hard coded for a single model type. Serve and Chat need to be able to handle any model vareity, namely `Q4_K_M` Mixtral variant to begin.

This issue is to track serve and config code that needs to change in order to handle this. 

relates to #851

### PR DESCRIPTION
**Pull request by cdoern**
both generate and serve already had a flag for the context size, meaning the user can override the default of 4096 in those two scenarios.

Added a new flag called `model-family` to serve which currently supports `merlinite` or `mixtral`. This is then used to load the proper chat template from a hardcoded map of supported templates. If nothing is specified, we default to the current one hardcoded in instructlab

Added the same flag to generate, since the generate prompt template needs extra padding for mixtral

Also, add `model-name` to `ilab train` as linux train had a hardcoded merlinite reference for the model name.

Most other "hardcoded" references are really just defaults which can be overridden by the user.
#973 speaks more about semantic fixes between serve and chat using the wrong model paths/names but is also meant to speak to hardcoded references to merlinite. #29 is the tracker for actually making sure serve uses the proper template.

resolves #973 
resolves #29

### CODE REVIEW

File: src/instructlab/generator/generate_data.py [hunk 1]
```diff
@@ -59,6 +59,27 @@
...
+* Strictly adhere to the prompt and generate responses in the same style and format as the example.
+Use this format to generate the questions: 
+### Question 1: 
+
+Now generate 5 questions following the example below, remember to follow the principles mentioned above and use the same format as the examples. Remember to use the same style and format as the example below. 
+
+To better assist you with this task, here is an example:
+### Question 1:
+"""
+
```
Comment (oindrillac) [inline src/instructlab/generator/generate_data.py:82]:
```suggestion
DEFAULT_PROMPT_TEMPLATE_MIXTRAL = """\
<s> [INST]You are a very knowledgeable AI Assistant that will faithfully assist the user with their task. You are asked to come up with a set of 5 diverse task instructions under {{taxonomy}}{{" for the task \\"%s\\""|format(task_description)  if task_description}}. These task instructions will be given to a GPT model and we will evaluate the GPT model for completing the instructions.

Here are the requirements:
1. Try not to repeat the verb for each instruction to maximize diversity.
2. The language used for the instruction also should be diverse. For example, you should combine questions with imperative instructions.
{% if not document -%}
3. The type of instructions should not have topic diversity. The list should follow the same topic and category.
{% else -%}
3. The type of instructions should be similar to provided examples. The generated instruction and the output should be grounded in the provided document.
{% endif -%}
4. A GPT language model should be able to complete the instruction. For example, do not ask the assistant to create any visual or audio output. For another example, do not ask the assistant to wake you up at 5pm or set a reminder because it cannot perform any action.
5. The instructions should be in English.
6. The instructions should be 1 to 2 sentences long. Either an imperative sentence or a question is permitted.
{% if not document -%}
7. You should generate an appropriate input to the instruction. The input field should contain a specific example provided for the instruction. It should involve realistic data and should not contain simple placeholders. The input should provide substantial content to make the instruction challenging but should ideally not exceed 100 words.
8. Not all instructions require input. For example, when an instruction asks about some general information, "what is the highest peak in the world", it is not necessary to provide a specific context. In this case, we simply put "<noinput>" in the input field.
9. The output should be an appropriate response to the instruction and the input. Make sure the output is less than 100 words.
{% else -%}
7. The output should be an appropriate response to the input and the instruction. Long outputs are preferable.
{% endif %}

{% if not document -%}
List of 5 tasks:
{% else -%}
Based on below document provide a list of 5 tasks:

Document:
{{document}}

Here are some examples to help you understand the type of questions that are asked for this document:
{% endif -%}[/INST]
"""

```

File: src/instructlab/lab.py [hunk 2]
```diff
@@ -685,16 +716,32 @@ def chat(
...
+            if not os.path.exists(os.path.join(model_dir, "safetensors")):
+                os.mkdir(os.path.join(model_dir, "safetensors"))
+            for f in files:
+                # if fullmodel we need to download all safetensors
+                hf_hub_download(
+                    token=os.getenv("HF_TOKEN"),
+                    repo_id=repository,
+                    revision=release,
+                    filename=f,
+                    local_dir=os.path.join(model_dir, "safetensors"),
```
Comment (jaideepr97) [inline src/instructlab/lab.py:735]: we will need to download more than just the safetensor files. we will also need the config.json for the model's metadata as well as all the tokenizer files

It would be better to download the whole repo IMO

Comment (cdoern) [reply src/instructlab/lab.py:735]: Sorry, I just named the dir safetenors. It downloads all the files in the repo

Comment (cdoern) [reply src/instructlab/lab.py]: moved to another branch

File: src/instructlab/train/linux_train.py [hunk 3]
```diff
@@ -107,6 +107,7 @@ def linux_train(
     num_epochs: Optional[int] = None,
     device: torch.device = torch.device("cpu"),
     four_bit_quant: bool = False,
+    model_name: str = "instructlab/merlinite-7b-lab"
```
[the remark this answers is not in the document]
Comment (khaledsulayman) [reply src/instructlab/train/linux_train.py]: this line is still here

**Comment by mairin** [comment]
need to run an end-to-end test.... run this with a granite model, ensure that it correctly loads the merlinite template and works correctly

File: src/instructlab/lab.py [hunk 4]
```diff
@@ -467,13 +487,15 @@ def generate(
         api_base = endpoint_url
     else:
         try:
+            # might need to funnel model-path into this?
```
Comment (cdoern) [inline src/instructlab/lab.py:490]: fixing this in next push

File: src/instructlab/lab.py [hunk 5]
```diff
@@ -688,13 +718,13 @@ def download(ctx, repository, release, filename, model_dir):
...
         if ctx.obj is not None:
             hf_logging.set_verbosity(ctx.obj.config.general.log_level.upper())
-
-        hf_hub_download(
-            repo_id=repository,
-            revision=release,
-            filename=filename,
-            local_dir=model_dir,
-        )
+            hf_hub_download(
```
Comment (cdoern) [inline src/instructlab/lab.py:721]: this is causing test failures, fixing in next push after reviewers.

Comment (cdoern) [reply src/instructlab/lab.py:721]: (indentation is the issue)

**Comment by cdoern** [comment]
The issue was `self.assertEqual(len(linux_train_mock.call_args[1]), 5)` since I added model_name to `train`. This was hardcoded to merlinite.

File: src/instructlab/lab.py [hunk 6]
```diff
@@ -688,8 +716,8 @@ def download(ctx, repository, release, filename, model_dir):
     try:
         if ctx.obj is not None:
             hf_logging.set_verbosity(ctx.obj.config.general.log_level.upper())
-
         hf_hub_download(
+            token=os.getenv("HF_TOKEN"),
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/lab.py:720]: thank you. This is left over from the `download` and `convert` code.

**Comment by cdoern** [comment]
done pushing for now. Feel free to review :)

File: src/instructlab/train/linux_train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/train/linux_train.py]: resolved. Thanks!

**Review by khaledsulayman — APPROVED** [review]

File: src/instructlab/lab.py [hunk 7]
```diff
@@ -828,6 +855,12 @@ def convert(self, value, param, ctx) -> "torch.device":
         "(reduces GPU VRAM usage and may slow down training)"
     ),
 )
+@click.option(
+    "--model-name",
+    default="instructlab/merlinite-7b-lab",
+    show_default=True,
+    help="model name to use in training",
```
Comment (alimaredia) [inline src/instructlab/lab.py:862]: chore: Might want to change this to something like  "huggingface path to use in training" just so users don't only put only `merlinite-7b-lab` instead of `instructlab/merlinite-7b-lab`.

File: src/instructlab/lab.py [hunk 8]
```diff
@@ -292,8 +292,14 @@ def diff(ctx, taxonomy_path, taxonomy_base, yaml_rules, quiet):
     type=click.INT,
     help="The context size is the maximum number of tokens considered by the model, for both the prompt and response. Defaults to 4096.",
 )
+@click.option(
+    "--model-family",
+    type=str,
+    default="merlinite",
+    help="Model family is used to specify which chat template to serve with",
+)
```
Comment (anik120) [inline src/instructlab/lab.py:300]: I'm really confused, why are chatting/serving with different model families? 
Why do we need a new flag? 
@oindrillac maybe you can help clarify? I was under the assumption that this PR will only have changes to `generate` based off of the last conversation.

File: src/instructlab/generator/generate_data.py [hunk 9]
```diff
@@ -302,6 +337,7 @@ def generate_data(
     logger,
     api_base,
     tls_insecure,
+    model_family: str,
```
Comment (anik120) [inline src/instructlab/generator/generate_data.py:340]: Do we need to introduce a new `model_family`? Can we not interpret the model family from `model_name`? 
Name: `merlinite-4b-asdf.gguf`. Family: `merlinite`
Name: `granite-4b-asdf.gguf`. Family: `merlinite` 
Name: `mixtral-4b-asdf.gguf`. Family: `mixtral` 

This solves for what Mo commented too, without any additional code changes.

**Review by anik120 — DISMISSED** [review]
@jaideepr97 @cdoern could you help clarify the intention of the PR with the changes to the title/descriptions you folks mentioned in the call please, it'll help with the review process, especially grokking failing test cases.

File: src/instructlab/lab.py (hunk 8 above)
Comment (alimaredia) [reply src/instructlab/lab.py:300]: `ilab generate` generates based on the model that's served by `ilab serve`. The model that's served via `ilab serve` is downloaded via `ilab download`. It's absolutely in scope to have `download`, `serve`, and `generate` parameterized to use a different model for `generate`.

File: src/instructlab/server.py [hunk 10]
```diff
@@ -155,12 +171,23 @@ def server(
...
 
+    template = ""
+    eos_token = "<|endoftext|>"
+    bos_token = ""
+    for template_type in templates:
+        if template_type["model"] == model_family:
+            template = template_type["template"]
+            if template == "mixtral":
+                eos_token = "</s>"
+                bos_token = "<s>"
```
Comment (jaideepr97) [inline src/instructlab/server.py:182]:
```suggestion
    for template_dict in templates:
        if template_dict["model"] == model_family:
            template = template_dict["template"]
            if template_dict["model"] == "mixtral":
                eos_token = "</s>"
                bos_token = "<s>"
```

Comment (cdoern) [reply src/instructlab/server.py:182]: Thanks. I caught this as well and will be fixing in my next push

File: src/instructlab/lab.py [hunk 11]
```diff
@@ -428,6 +435,16 @@ def serve(ctx, model_path, gpu_layers, num_threads, max_ctx_size):
     default="",
     help="TLS client certificate password.",
 )
+@click.option(
+    "--model-path",
+    default=None,
+    help="model path to override the server config's model_path",
+)
```
Comment (jaideepr97) [inline src/instructlab/lab.py:442]: I dont think this change is required at the moment

File: src/instructlab/lab.py [hunk 12]
```diff
@@ -474,6 +493,7 @@ def generate(
                 tls_client_cert,
                 tls_client_key,
                 tls_client_passwd,
+                model_path,
```
Comment (jaideepr97) [inline src/instructlab/lab.py:496]: same as https://github.com/instructlab/instructlab/pull/1002/files#r1583411423

File: src/instructlab/lab.py [hunk 13]
```diff
@@ -448,6 +465,8 @@ def generate(
     tls_client_cert,
     tls_client_key,
     tls_client_passwd,
+    model_path,
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/lab.py:468]: ack, sure. This can be separate PR.

File: src/instructlab/lab.py [hunk 14]
```diff
@@ -526,7 +547,12 @@ def generate(
 @click.option(
     "-m",
     "--model",
-    help="Model to use",
+    help="Model name to print in chat process",
+)
+@click.option(
+    "--model-path",
+    default=None,
+    help="Model path to use when starting server",
```
Comment (jaideepr97) [inline src/instructlab/lab.py:555]: same as https://github.com/instructlab/instructlab/pull/1002/files#r1583411423

Comment (cdoern) [reply src/instructlab/lab.py:555]: ack, sure. This can be separate PR.

File: src/instructlab/server.py [hunk 15]
```diff
@@ -155,12 +171,23 @@ def server(
             return
         raise ServerException(f"failed creating the server application: {exc}") from exc
 
+    template = ""
+    eos_token = "<|endoftext|>"[z]
```
Comment (jaideepr97) [inline src/instructlab/server.py:175]: remove extraneous [z]

**Comment by jaideepr97** [comment]
@anik120 updated the PR title

The changes included in this PR do the following:
- changes to `serve` so that mixtral chat template is passed to the model server
- changes to `generate` so that mixtral generation template is passed to the model server 
- changes to `train` so that any model other than merlinite can be trained on linux

File: src/instructlab/generator/generate_data.py [hunk 16]
```diff
@@ -81,14 +111,19 @@
...
     except FileNotFoundError:
-        print(f"Cannot find {prompt_file_path}. Using default prompt.")
-        prompt_template = DEFAULT_PROMPT_TEMPLATE
+        print(
+            f"Cannot find {prompt_file_path}. Using default prompt depending on model-family."
+        )
+        if model_family == "merlinite":
+            prompt_template = DEFAULT_PROMPT_TEMPLATE_MERLINITE
+        elif model_family == "mixtral":
+            prompt_template = DEFAULT_PROMPT_TEMPLATE_MIXTRAL
```
Comment (tiran) [inline src/instructlab/generator/generate_data.py:126]: Include an error case so an unknown family does not cause a NameError:
```suggestion
            prompt_template = DEFAULT_PROMPT_TEMPLATE_MIXTRAL
        else:
            raise ValueError(f"Unsupported family '{model_family}'")
```

File: src/instructlab/generator/generate_data.py [hunk 17]
```diff
@@ -81,14 +111,21 @@
...
     """Check for prompt file."""
     try:
         with open(prompt_file_path, encoding="utf=8") as file:
             prompt_template = file.read()
     except FileNotFoundError:
-        print(f"Cannot find {prompt_file_path}. Using default prompt.")
-        prompt_template = DEFAULT_PROMPT_TEMPLATE
+        print(
+            f"Cannot find {prompt_file_path}. Using default prompt depending on model-family."
+        )
```
Comment (nathan-weinberg) [inline src/instructlab/generator/generate_data.py:122]: Any reason we are using `print` here instead of the logger?

File: src/instructlab/generator/generate_data.py (hunk 9 above)
Comment (nathan-weinberg) [reply src/instructlab/generator/generate_data.py:342]: Was there a resolution to this? I'm curious about this point as well

File: src/instructlab/lab.py [hunk 18]
```diff
@@ -296,8 +296,14 @@ def diff(ctx, taxonomy_path, taxonomy_base, yaml_rules, quiet):
     type=click.INT,
     help="The context size is the maximum number of tokens considered by the model, for both the prompt and response. Defaults to 4096.",
 )
+@click.option(
+    "--model-family",
+    type=str,
+    default="merlinite",
+    help="Model family is used to specify which chat template to serve with",
```
Comment (nathan-weinberg) [inline src/instructlab/lab.py:303]:
```suggestion
    help="Model family chat template to serve with, e.g. 'merlinite', 'granite', etc.",
```

File: src/instructlab/lab.py [hunk 19]
```diff
@@ -432,6 +439,11 @@ def serve(ctx, model_path, gpu_layers, num_threads, max_ctx_size):
     default="",
     help="TLS client certificate password.",
 )
+@click.option(
+    "--model-family",
+    default="merlinite",
+    help="model family to use when picking a generation template",
```
Comment (nathan-weinberg) [inline src/instructlab/lab.py:445]:
```suggestion
    help="Model family generation template to serve with, e.g. 'merlinite', 'granite', etc.",
```

File: src/instructlab/lab.py [hunk 20]
```diff
@@ -593,6 +607,12 @@ def generate(
     default="",
     help="TLS client certificate password.",
 )
+@click.option(
+    "--model-family",
+    default="merlinite",
+    show_default=True,
+    help="model family to use when picking a chat template",
```
Comment (nathan-weinberg) [inline src/instructlab/lab.py:614]:
```suggestion
    help="Model family to use when picking a chat template, e.g. 'merlinite', 'granite', etc.",
```

File: src/instructlab/lab.py (hunk 7 above)
Comment (nathan-weinberg) [reply src/instructlab/lab.py:860]: Can we extrapolate any of this data from other args? If not let's update the help text so it's less confusing for users

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]

File: src/instructlab/generator/generate_data.py (hunk 17 above)
Comment (cdoern) [reply src/instructlab/generator/generate_data.py:122]: This was always a print, I just changed what was being printed. Opting to not change logic in this PR if I don't have to.

Comment (nathan-weinberg) [reply src/instructlab/generator/generate_data.py:122]: fair enough!

File: src/instructlab/lab.py (hunk 20 above)
Comment (jaideepr97) [reply src/instructlab/lab.py:615]: I think this is a good thing to point out - though I would argue we can tackle this in a follow up PR. There is some ambiguity in how this help message is written because this parameter decides 2 different things. 
- which template to use for `generate`
- also which template to use for `chat` if a server is started in the background by `generate`

so we can think about structuring it a bit better and handle it in a follow up PR

**Review by russellb — APPROVED** [review]
I think this is good and we can continue with follow-up PRs to improve UX, help text, and so forth - I'd like to go ahead and merge since some others are waiting for the functional part to be merged

**Review by nathan-weinberg — APPROVED** [review]
Comments will be addressed in followup

File: src/instructlab/lab.py (hunk 20 above)
Comment (nathan-weinberg) [reply src/instructlab/lab.py:615]: Right now it's very ambiguous to users, so definitely needs to be followed up on

File: src/instructlab/generator/generate_data.py (hunk 9 above)
Comment (anik120) [reply src/instructlab/generator/generate_data.py:342]: Nah I was told ignore for this PR it'll be a follow up 🤷🏽

**Comment by bbrowning** [comment]
Should we be using the chat templates as specified in the `tokenizer_config.json` of each model from HuggingFace instead of hardcoding them? Essentially https://huggingface.co/docs/transformers/main/en/chat_templating#how-do-i-use-chat-templates

This would get us out of having to hardcode the templates at all and instead we'll just use what the model's metadata says it requires. This would mean expecting models to maintain appropriate chat templates, but I believe that's a pretty common practice.

**Comment by russellb** [comment]
Definitely sounds like a good follow-up issue to look at! Thanks, @bbrowning !
